runtime/cudaq/utils/extension_point.h

```diff
@@ -0,0 +1,168 @@
+///
+/// How to use the extension_point class
+///
+/// The extension_point class provides a mechanism for creating extensible
+/// frameworks with plugin-like functionality. Here's how to use it:
+///
+/// 1. Define your extension point:
+///    Create a new class that inherits from cudaq::extension_point<YourClass>.
+///    This class should declare pure virtual methods that extensions will
+///    implement.
+///
+/// @code
+/// class MyExtensionPoint : public cudaq::extension_point<MyExtensionPoint> {
+/// public:
+///   virtual std::string parrotBack(const std::string &msg) const = 0;
+/// };
+/// @endcode
+///
+/// 2. Implement concrete extensions:
+///    Create classes that inherit from your extension point and implement its
+///    methods. Use the CUDAQ_EXTENSION_CREATOR_FUNCTION macro to define a
+///    creator function.
+///
+/// @code
+/// class RepeatBackOne : public MyExtensionPoint {
+/// public:
+///   std::string parrotBack(const std::string &msg) const override {
+///     return msg + " from RepeatBackOne.";
+///   }
+///
+///   CUDAQ_EXTENSION_CREATOR_FUNCTION(MyExtensionPoint, RepeatBackOne)
+/// };
+/// @endcode
+///
+/// 3. Register your extensions:
+///    Use the CUDAQ_REGISTER_TYPE macro to register each extension.
+///
+/// @code
+/// CUDAQ_REGISTER_TYPE(RepeatBackOne)
+/// @endcode
+///
+/// 4. Use your extensions:
+///    You can now create instances of your extensions, check registrations, and
+///    more.
+///
+/// @code
+/// auto extension = MyExtensionPoint::get("RepeatBackOne");
+/// std::cout << extension->parrotBack("Hello") << std::endl;
+///
+/// auto registeredTypes = MyExtensionPoint::get_registered();
+/// bool isRegistered = MyExtensionPoint::is_registered("RepeatBackOne");
+/// @endcode
+///
+/// This approach allows for a flexible, extensible design where new
+/// functionality can be added without modifying existing code.
+template <typename T, typename... CtorArgs>
+class extension_point {
+
+  /// Type alias for the creator function.
+  using CreatorFunction = std::function<std::unique_ptr<T>(CtorArgs...)>;
+
+protected:
+  /// @brief Get the registry of creator functions.
+  /// @return A reference to the static registry map.
+  static std::unordered_map<std::string, CreatorFunction> &get_registry() {
+    static std::unordered_map<std::string, CreatorFunction> registry;
+    return registry;
+  }
+
+public:
+  /// @brief Create an instance of a registered extension.
+  /// @param name The identifier of the registered extension.
+  /// @param args Constructor arguments for the extension.
+  /// @return A unique pointer to the created instance.
+  /// @throws std::runtime_error if the extension is not found.
+  static std::unique_ptr<T> get(const std::string &name, CtorArgs... args) {
+    auto &registry = get_registry();
+    auto iter = registry.find(name);
+    if (iter == registry.end())
+      throw std::runtime_error("Cannot find extension with name = " + name);
```

Comment:
I think nvcc needed this change. The bottom line is that it would be good for somebody to add a test with a .cu file to make sure all of this works with the nvcc compiler.
```suggestion
      throw std::runtime_error(
          std::string("Cannot find extension with name = ") + name);
```